VSRAD.Syntax/Options/BaseOptionModel.cs

```diff
@@ -0,0 +1,154 @@
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Shell.Settings;
+using Microsoft.VisualStudio.Threading;
+using VSRAD.Syntax.Helpers;
+using Task = System.Threading.Tasks.Task;
+
+namespace VSRAD.Syntax.Options
+{
+    public abstract class BaseOptionModel<T> where T : BaseOptionModel<T>, new()
+    {
+        private static readonly AsyncLazy<T> _liveModel = new AsyncLazy<T>(CreateAsync, ThreadHelper.JoinableTaskFactory);
+        protected static readonly AsyncLazy<ShellSettingsManager> _settingsManager = new AsyncLazy<ShellSettingsManager>(GetSettingsManagerAsync, ThreadHelper.JoinableTaskFactory);
+
+        protected BaseOptionModel()
+        { }
+
+        public static T Instance
+        {
+            get
+            {
+                ThreadHelper.ThrowIfNotOnUIThread();
+                return ThreadHelper.JoinableTaskFactory.Run(GetInstanceAsync);
+            }
+        }
+
+        public static Task<T> GetInstanceAsync() => _liveModel.GetValueAsync();
+
+        public static async Task<T> CreateAsync()
+        {
+            var instance = new T();
+            await instance.LoadAsync();
+            return instance;
+        }
+
+        protected virtual string CollectionName { get; } = typeof(T).FullName;
+
+        public virtual void Load() => ThreadHelper.JoinableTaskFactory.Run(LoadAsync);
+
+        public virtual async Task LoadAsync()
+        {
+            var manager = await _settingsManager.GetValueAsync();
+            var settingsStore = manager.GetReadOnlySettingsStore(SettingsScope.UserSettings);
+
+            if (!settingsStore.CollectionExists(CollectionName))
+                return;
+
+            foreach (PropertyInfo property in GetOptionProperties())
+            {
+                try
+                {
+                    var serializedProp = settingsStore.GetString(CollectionName, property.Name);
+                    var value = DeserializeValue(serializedProp, property.PropertyType);
+                    property.SetValue(this, value);
+                }
+                catch (Exception ex)
+                {
+                    Error.LogError(ex);
+                }
+            }
+        }
+
+        public virtual void Save() => ThreadHelper.JoinableTaskFactory.Run(SaveAsync);
+
+        public virtual async Task SaveAsync()
+        {
+            ShellSettingsManager manager = await _settingsManager.GetValueAsync();
+            WritableSettingsStore settingsStore = manager.GetWritableSettingsStore(SettingsScope.UserSettings);
+
+            if (!settingsStore.CollectionExists(CollectionName))
+                settingsStore.CreateCollection(CollectionName);
+
+            foreach (PropertyInfo property in GetOptionProperties())
+            {
+                string output = SerializeValue(property.GetValue(this));
+                settingsStore.SetString(CollectionName, property.Name, output);
```

Comment:
Code style consistency: it is better to use `var` when possible, or always use explicit type names

Reply:
Thank you, fixed in 024da88